client/cli/src/lib.rs

```diff
@@ -177,461 +179,393 @@ fn is_node_name_valid(_name: &str) -> Result<(), &str> {
-			ParseAndPrepareRun { params, impl_name, version }
-		),
-		params::CoreParams::BuildSpec(params) => ParseAndPrepare::BuildSpec(
-			ParseAndPrepareBuildSpec { params, version }
-		),
-		params::CoreParams::ExportBlocks(params) => ParseAndPrepare::ExportBlocks(
-			ParseAndPrepareExport { params, version }
-		),
-		params::CoreParams::ImportBlocks(params) => ParseAndPrepare::ImportBlocks(
-			ParseAndPrepareImport { params, version }
-		),
-		params::CoreParams::CheckBlock(params) => ParseAndPrepare::CheckBlock(
-			CheckBlock { params, version }
-		),
-		params::CoreParams::PurgeChain(params) => ParseAndPrepare::PurgeChain(
-			ParseAndPreparePurge { params, version }
-		),
-		params::CoreParams::Revert(params) => ParseAndPrepare::RevertChain(
-			ParseAndPrepareRevert { params, version }
-		),
-		params::CoreParams::Custom(params) => ParseAndPrepare::CustomCommand(params),
-	};
-	init_logger(args.shared_params().and_then(|p| p.log.as_ref()).map(|v| v.as_ref()).unwrap_or(""));
-	args
+	T::from_clap(&app.get_matches_from(iter))
 }
 
-/// Returns a string displaying the node role, special casing the sentry mode
-/// (returning `SENTRY`), since the node technically has an `AUTHORITY` role but
-/// doesn't participate.
-pub fn display_role<A, B, C>(config: &Configuration<A, B, C>) -> String {
-	if config.sentry_mode {
-		"SENTRY".to_string()
-	} else {
-		format!("{:?}", config.roles)
-	}
-}
-
-/// Output of calling `parse_and_prepare`.
-#[must_use]
-pub enum ParseAndPrepare<'a, CC, RP> {
-	/// Command ready to run the main client.
-	Run(ParseAndPrepareRun<'a, RP>),
-	/// Command ready to build chain specs.
-	BuildSpec(ParseAndPrepareBuildSpec<'a>),
-	/// Command ready to export the chain.
-	ExportBlocks(ParseAndPrepareExport<'a>),
-	/// Command ready to import the chain.
-	ImportBlocks(ParseAndPrepareImport<'a>),
-	/// Command to check a block.
-	CheckBlock(CheckBlock<'a>),
-	/// Command ready to purge the chain.
-	PurgeChain(ParseAndPreparePurge<'a>),
-	/// Command ready to revert the chain.
-	RevertChain(ParseAndPrepareRevert<'a>),
-	/// An additional custom command passed to `parse_and_prepare`.
-	CustomCommand(CC),
-}
+pub fn run<F, G, E, F2, F3, F4, T1, T2, T3, T5>(
+	core_params: CoreParams,
+	new_light: F2,
+	new_full: F3,
+	spec_factory: F,
+	builder: F4,
+	impl_name: &'static str,
+	version: &VersionInfo,
+) -> error::Result<()>
+where
+	F: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
+	F2: FnOnce(Configuration<G, E>) -> Result<T1, sc_service::error::Error>,
+	F3: FnOnce(Configuration<G, E>) -> Result<T2, sc_service::error::Error>,
+	F4: FnOnce(Configuration<G, E>) -> Result<T3, sc_service::error::Error>,
+	G: RuntimeGenesis,
+	E: ChainSpecExtension,
+	T1: AbstractService + std::marker::Unpin,
+	T2: AbstractService + std::marker::Unpin,
+	T3: ServiceBuilderCommand<Block = T5> + std::marker::Unpin,
+	T5: sp_runtime::traits::Block + Debug,
+	<<<T5 as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number as
```

Comment:
`<<<<T5 as BlockT>::Header as HeaderT>::Number as std::str::FromStr>::Error: std::fmt::Debug` is probably cleaner